src/bucket/BucketIndexUtils.h

```diff
@@ -0,0 +1,110 @@
+#pragma once
+
+// Copyright 2025 Stellar Development Foundation and contributors. Licensed
+// under the Apache License, Version 2.0. See the COPYING file at the root
+// of this distribution or at http://www.apache.org/licenses/LICENSE-2.0
+
+#include "util/GlobalChecks.h"
+#include "util/XDROperators.h" // IWYU pragma: keep
+#include "xdr/Stellar-ledger-entries.h"
+#include <filesystem>
+#include <map>
+#include <optional>
+#include <variant>
+#include <vector>
+
+#include "util/XDRCereal.h"
+#include <cereal/archives/binary.hpp>
+#include <cereal/types/utility.hpp>
+#include <cereal/types/vector.hpp>
+
+namespace asio
+{
+class io_context;
+}
+
+namespace stellar
+{
+
+class BucketManager;
+class Config;
+
+using AssetPoolIDMap = std::map<Asset, std::vector<PoolID>>;
+using IndexPtrT = std::shared_ptr<BucketEntry const>;
+
+// Querying a BucketIndex can return one of three states:
+// 1. CACHE_HIT: The entry is in the cache. Can either be a live or dead entry.
+// 2. FILE_OFFSET: The entry is not in the cache, but the entry potentially
+//    exists at the given offset.
+// 3. NOT_FOUND: The entry does not exist in the bucket.
+enum IndexReturnState
+{
+    CACHE_HIT,
+    FILE_OFFSET,
+    NOT_FOUND
+};
+
+class IndexReturnT
+{
+  private:
+    // Payload maps to the possible return states:
+    // CACHE_HIT: IndexPtrT
+    // FILE_OFFSET: std::streamoff
+    // NOT_FOUND: nullopt
+    using PayloadT = std::optional<std::variant<IndexPtrT, std::streamoff>>;
+    PayloadT mPayload;
+    IndexReturnState mState;
+
+  public:
+    IndexReturnT(IndexPtrT entry)
+        : mPayload(entry), mState(IndexReturnState::CACHE_HIT)
+    {
+        releaseAssertOrThrow(entry);
+    }
+    IndexReturnT(std::streamoff offset)
+        : mPayload(offset), mState(IndexReturnState::FILE_OFFSET)
+    {
+    }
+    IndexReturnT() : mState(IndexReturnState::NOT_FOUND)
+    {
+    }
+
+    IndexReturnState
+    getState() const
+    {
+        return mState;
+    }
+    IndexPtrT
+    cacheHit() const
+    {
+        releaseAssertOrThrow(mState == IndexReturnState::CACHE_HIT);
+        releaseAssertOrThrow(mPayload);
```

Comment:
nit: The assertion should probably be on `std::holds_alternative` instead.

Reply:
I think I'll keep both, just because the value of `mState` is how the user interacts with the wrapper class.